servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/EmptySubscriptions.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright © 2018, 2021 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.concurrent.internal;
+
+import io.servicetalk.concurrent.PublisherSource.Subscriber;
+import io.servicetalk.concurrent.PublisherSource.Subscription;
+
+import static io.servicetalk.concurrent.internal.SubscriberUtils.isRequestNValid;
+import static io.servicetalk.concurrent.internal.SubscriberUtils.newExceptionForInvalidRequestN;
+
+/**
+ * Utility methods for {@link Subscription}s which don't deliver any data.
+ */
+public final class EmptySubscriptions {
+    public static final Subscription EMPTY_SUBSCRIPTION = newEmptySubscription(true);
+    public static final Subscription EMPTY_SUBSCRIPTION_NO_THROW = newEmptySubscription(false);
+
+    private EmptySubscriptions() {
+    }
+
+    /**
+     * Create an empty {@link Subscription} that will propagate an error to a {@link Subscriber} upon invalid demand.
+     * @param subscriber The subscriber to propagate
+     * @param <T> The type of {@link Subscriber}.
+     * @return An empty {@link Subscription} that will propagate an error to a {@link Subscriber} upon invalid demand.
+     */
+    public static <T> Subscription newEmptySubscription(Subscriber<T> subscriber) {
+        return new Subscription() {
+            private boolean terminated;
+            @Override
+            public void request(final long n) {
+                if (!terminated && !isRequestNValid(n)) {
```

Comment:
The [rule 3.9](https://github.com/reactive-streams/reactive-streams-jvm#3.9) says:

> While the `Subscription` is not cancelled, `Subscription.request(long n)` MUST signal `onError` ...

We need to mark it as `terminated` inside `cancel()`. That also helps to satisfy [rule 3.6](https://github.com/reactive-streams/reactive-streams-jvm#3.6):

> After the `Subscription` is cancelled, additional `Subscription.request(long n)` MUST be NOPs.

Reply:
humm good point. I think the important part is we don't terminate the subscriber more than once, seems strange to "suppress" exceptions just because cancel had been previously called, but it can be done simple enough.